# LightAutoML (LAMA) Baselines

В данном ноутбуке представлены базовые модели на основе библиотеки LightAutoML (LAMA)

**Цель**: построить baseline-модели для задачи регрессии (предсказание страховых выплат) и сравнить различные конфигурации LAMA.

Log MAE — средняя абсолютная ошибка на логарифмированном таргете (официальная метрика соревнования Allstate Claims Severity).


In [1]:
import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from src.data.loader import DataLoader
from src.features.eda_insights import load_eda_insights
from src.models.baseline import LAMABaseline
from src.utils.logging_config import setup_logging, get_logger
from src.utils.metrics import compute_log_mae, compute_metrics

warnings.filterwarnings('ignore')
setup_logging()
logger = get_logger(__name__)

plt.style.use('seaborn-v0_8-whitegrid')
sns.set_palette('husl')
%matplotlib inline

In [2]:
loader = DataLoader()
train_df = loader.load_train()
test_df = loader.load_test()

2025-12-20 21:10:28,473 - src.data.loader - INFO - Loaded 188,318 training samples
2025-12-20 21:10:28,858 - src.data.loader - INFO - Loaded 125,546 test samples


## Конфигурация 1 (default)

Базовая конфигурация LAMA использует простой одноуровневый ансамбль с блендингом:

**Алгоритмы:**
- `lgb` — LightGBM (градиентный бустинг)
- `linear_l2` — Ridge-регрессия (линейная модель с L2-регуляризацией)

**Особенности:**
- 5-fold кросс-валидация
- Финальное предсказание: взвешенная комбинация моделей (blending)
- Используются алгоритмы с дефолтными гиперпараметрами (без тюнинга)
- Веса блендинга оптимизируются автоматически на основе CV-скоров

In [3]:
eda_insights = load_eda_insights()

lama_default = LAMABaseline(
    config_name="default",
    timeout=600,
    n_threads=4,
    n_folds=5,
    random_state=42,
    eda_insights=eda_insights,
)

lama_default.fit(train_df, target_col="loss", drop_cols=["id"])

y_true = train_df["loss"].values
metrics_default = lama_default.evaluate(y_true, return_all_metrics=True)

print(f"Log MAE: {metrics_default['log_mae']:.4f}")


2025-12-20 21:10:28,862 - src.models.baseline - INFO - Fitting LAMA baseline (default), shape: (188318, 132)
[21:10:28] Stdout logging level is INFO.
[21:10:28] Copying TaskTimer may affect the parent PipelineTimer, so copy will create new unlimited TaskTimer
[21:10:28] Task: reg

[21:10:28] Start automl preset with listed constraints:
[21:10:28] - time: 600.00 seconds
[21:10:28] - CPU: 4 cores
[21:10:28] - memory: 16 GB

[21:10:28] Train data shape: (188318, 132)

[21:10:34] Layer 1 train process start. Time left 594.24 secs
[21:10:38] Start fitting Lvl_0_Pipe_0_Mod_0_LinearL2 ...
[21:11:07] Fitting Lvl_0_Pipe_0_Mod_0_LinearL2 finished. score = -1302.0093579111804
[21:11:07] Lvl_0_Pipe_0_Mod_0_LinearL2 fitting and predicting completed
[21:11:07] Time left 561.54 secs

[21:11:10] Selector_LightGBM fitting and predicting completed
[21:11:14] Start fitting Lvl_0_Pipe_1_Mod_0_LightGBM ...
[21:11:34] Fitting Lvl_0_Pipe_1_Mod_0_LightGBM finished. score = -1909.5202468069765
[21:11:34] Lvl_0

## Конфигурация 2 (tuned)

Улучшенная конфигурация LAMA использует многоуровневую архитектуру

**Архитектура:**
- **Уровень 1:** `linear_l2` + `lgb` — базовые модели
- **Уровень 2:** `lgb_tuned` — LightGBM с оптимизацией гиперпараметров через Optuna

**Особенности:**
- 5-fold кросс-валидация
- До 100 итераций оптимизации гиперпараметров
- Модель второго уровня обучается на предсказаниях моделей первого уровня
- `TabularUtilizedAutoML` запускает несколько конфигураций с разными `random_state` и объединяет результаты


In [4]:
lama_tuned = LAMABaseline(
    config_name="tuned",
    timeout=900,
    n_threads=4,
    n_folds=5,
    random_state=42,
    eda_insights=eda_insights,
)

lama_tuned.fit(train_df, target_col="loss", drop_cols=["id"])

metrics_tuned = lama_tuned.evaluate(y_true, return_all_metrics=True)

print(f"Log MAE: {metrics_tuned['log_mae']:.4f}")


2025-12-20 21:11:34,553 - src.models.baseline - INFO - Fitting LAMA baseline (tuned), shape: (188318, 132)
[21:11:34] Start automl utilizator with listed constraints:
[21:11:34] - time: 900.00 seconds
[21:11:34] - CPU: 4 cores
[21:11:34] - memory: 16 GB

[21:11:34] If one preset completes earlier, next preset configuration will be started

[21:11:34] ==================================================
[21:11:34] Start 0 automl preset configuration:
[21:11:34] conf_0_sel_type_0.yml, random state: {'reader_params': {'random_state': 42}, 'nn_params': {'random_state': 42}, 'general_params': {'return_all_predictions': False}}
[21:11:34] Stdout logging level is INFO.
[21:11:34] Task: reg

[21:11:34] Start automl preset with listed constraints:
[21:11:34] - time: 900.00 seconds
[21:11:34] - CPU: 4 cores
[21:11:34] - memory: 16 GB

[21:11:34] Train data shape: (188318, 132)

[21:11:40] Layer 1 train process start. Time left 894.19 secs
[21:11:44] Start fitting Lvl_0_Pipe_0_Mod_0_LinearL2 ...
[2

Optimization Progress:   0%|          | 0/100 [00:00<?, ?it/s]

2025-12-20 21:13:03,111 - optuna.storages._in_memory - INFO - A new study created in memory with name: no-name-b5cab15a-c938-4445-b9a1-a46dd4ec930c
2025-12-20 21:13:13,272 - optuna.study.study - INFO - Trial 0 finished with value: -1187.48341658606 and parameters: {'feature_fraction': 0.6872700594236812, 'num_leaves': 244, 'bagging_fraction': 0.8659969709057025, 'min_sum_hessian_in_leaf': 0.24810409748678114}. Best is trial 0 with value: -1187.48341658606.


Optimization Progress:   1%|          | 1/100 [00:10<16:46, 10.17s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:13:18,034 - optuna.study.study - INFO - Trial 1 finished with value: -1191.0728282674693 and parameters: {'feature_fraction': 0.5780093202212182, 'num_leaves': 53, 'bagging_fraction': 0.5290418060840998, 'min_sum_hessian_in_leaf': 2.915443189153755}. Best is trial 0 with value: -1187.48341658606.


Optimization Progress:   2%|▏         | 2/100 [00:14<11:24,  6.99s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:13:23,797 - optuna.study.study - INFO - Trial 2 finished with value: -1191.3287121607311 and parameters: {'feature_fraction': 0.8005575058716043, 'num_leaves': 185, 'bagging_fraction': 0.5102922471479012, 'min_sum_hessian_in_leaf': 7.5794799533480015}. Best is trial 0 with value: -1187.48341658606.


Optimization Progress:   3%|▎         | 3/100 [00:20<10:23,  6.43s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:13:27,974 - optuna.study.study - INFO - Trial 3 finished with value: -1191.2597674304182 and parameters: {'feature_fraction': 0.9162213204002109, 'num_leaves': 66, 'bagging_fraction': 0.5909124836035503, 'min_sum_hessian_in_leaf': 0.00541524411940254}. Best is trial 0 with value: -1187.48341658606.


Optimization Progress:   4%|▍         | 4/100 [00:24<08:51,  5.54s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:13:35,592 - optuna.study.study - INFO - Trial 4 finished with value: -1186.38869012656 and parameters: {'feature_fraction': 0.6521211214797689, 'num_leaves': 141, 'bagging_fraction': 0.7159725093210578, 'min_sum_hessian_in_leaf': 0.014618962793704969}. Best is trial 4 with value: -1186.38869012656.


Optimization Progress:   5%|▌         | 5/100 [00:32<09:57,  6.29s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:13:39,376 - optuna.study.study - INFO - Trial 5 finished with value: -1189.9709886931885 and parameters: {'feature_fraction': 0.8059264473611898, 'num_leaves': 49, 'bagging_fraction': 0.6460723242676091, 'min_sum_hessian_in_leaf': 0.029204338471814112}. Best is trial 4 with value: -1186.38869012656.


Optimization Progress:   6%|▌         | 6/100 [00:36<08:31,  5.44s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:13:48,371 - optuna.study.study - INFO - Trial 6 finished with value: -1189.0475877673523 and parameters: {'feature_fraction': 0.728034992108518, 'num_leaves': 204, 'bagging_fraction': 0.5998368910791798, 'min_sum_hessian_in_leaf': 0.11400863701127326}. Best is trial 4 with value: -1186.38869012656.


Optimization Progress:   7%|▋         | 7/100 [00:45<10:13,  6.60s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:13:54,623 - optuna.study.study - INFO - Trial 7 finished with value: -1187.1484791078299 and parameters: {'feature_fraction': 0.7962072844310213, 'num_leaves': 27, 'bagging_fraction': 0.8037724259507192, 'min_sum_hessian_in_leaf': 0.004809461967501573}. Best is trial 4 with value: -1186.38869012656.


Optimization Progress:   8%|▊         | 8/100 [00:51<09:57,  6.49s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:14:09,045 - optuna.study.study - INFO - Trial 8 finished with value: -1186.3937809350557 and parameters: {'feature_fraction': 0.5325257964926398, 'num_leaves': 243, 'bagging_fraction': 0.9828160165372797, 'min_sum_hessian_in_leaf': 1.7123375973163983}. Best is trial 4 with value: -1186.38869012656.


Optimization Progress:   9%|▉         | 9/100 [01:05<13:36,  8.97s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:14:17,763 - optuna.study.study - INFO - Trial 9 finished with value: -1185.7586989847277 and parameters: {'feature_fraction': 0.6523068845866853, 'num_leaves': 39, 'bagging_fraction': 0.8421165132560784, 'min_sum_hessian_in_leaf': 0.057624872164786005}. Best is trial 9 with value: -1185.7586989847277.


Optimization Progress:  10%|█         | 10/100 [01:14<13:20,  8.89s/it, best_trial=9, best_value=-1.19e+3]

2025-12-20 21:14:23,642 - optuna.study.study - INFO - Trial 10 finished with value: -1190.0418385736275 and parameters: {'feature_fraction': 0.9725682721151933, 'num_leaves': 103, 'bagging_fraction': 0.9267360518846169, 'min_sum_hessian_in_leaf': 0.0011799062523159282}. Best is trial 9 with value: -1185.7586989847277.


Optimization Progress:  11%|█         | 11/100 [01:20<11:49,  7.97s/it, best_trial=9, best_value=-1.19e+3]

2025-12-20 21:14:33,698 - optuna.study.study - INFO - Trial 11 finished with value: -1187.222015571189 and parameters: {'feature_fraction': 0.6075747423853284, 'num_leaves': 135, 'bagging_fraction': 0.7167565640612156, 'min_sum_hessian_in_leaf': 0.02616865856724872}. Best is trial 9 with value: -1185.7586989847277.


Optimization Progress:  12%|█▏        | 12/100 [01:30<12:37,  8.60s/it, best_trial=9, best_value=-1.19e+3]

2025-12-20 21:14:40,427 - optuna.study.study - INFO - Trial 12 finished with value: -1185.5412967928528 and parameters: {'feature_fraction': 0.6531277350796226, 'num_leaves': 107, 'bagging_fraction': 0.7550056138591037, 'min_sum_hessian_in_leaf': 0.2949971987301987}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  13%|█▎        | 13/100 [01:37<11:39,  8.04s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:14:51,179 - optuna.study.study - INFO - Trial 13 finished with value: -1185.7069655800674 and parameters: {'feature_fraction': 0.5207781426982558, 'num_leaves': 92, 'bagging_fraction': 0.8103570903814009, 'min_sum_hessian_in_leaf': 0.4796131572545122}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  14%|█▍        | 14/100 [01:48<12:41,  8.86s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:01,769 - optuna.study.study - INFO - Trial 14 finished with value: -1186.187082068833 and parameters: {'feature_fraction': 0.5098079084751951, 'num_leaves': 93, 'bagging_fraction': 0.7793108233176906, 'min_sum_hessian_in_leaf': 0.4309974718367054}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  15%|█▌        | 15/100 [01:58<13:17,  9.38s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:11,707 - optuna.study.study - INFO - Trial 15 finished with value: -1187.1871790502996 and parameters: {'feature_fraction': 0.5834550402472981, 'num_leaves': 128, 'bagging_fraction': 0.8892016965842912, 'min_sum_hessian_in_leaf': 0.6630827574066053}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  16%|█▌        | 16/100 [02:08<13:21,  9.55s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:20,787 - optuna.study.study - INFO - Trial 16 finished with value: -1187.554970895545 and parameters: {'feature_fraction': 0.5085996798089254, 'num_leaves': 82, 'bagging_fraction': 0.7406442464830113, 'min_sum_hessian_in_leaf': 0.16317854542245658}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  17%|█▋        | 17/100 [02:17<13:00,  9.41s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:28,186 - optuna.study.study - INFO - Trial 17 finished with value: -1188.0570525636083 and parameters: {'feature_fraction': 0.7133656678357854, 'num_leaves': 162, 'bagging_fraction': 0.7039294159774774, 'min_sum_hessian_in_leaf': 1.1660348298902208}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  18%|█▊        | 18/100 [02:25<12:01,  8.80s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:35,306 - optuna.study.study - INFO - Trial 18 finished with value: -1186.12707942201 and parameters: {'feature_fraction': 0.622344146322471, 'num_leaves': 106, 'bagging_fraction': 0.8169167791036113, 'min_sum_hessian_in_leaf': 9.602051763701697}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  19%|█▉        | 19/100 [02:32<11:12,  8.30s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:43,792 - optuna.study.study - INFO - Trial 19 finished with value: -1186.9674162975368 and parameters: {'feature_fraction': 0.5506262176536811, 'num_leaves': 77, 'bagging_fraction': 0.6603420358958465, 'min_sum_hessian_in_leaf': 0.4138444779876123}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  20%|██        | 20/100 [02:40<11:08,  8.35s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:48,788 - optuna.study.study - INFO - Trial 20 finished with value: -1188.2434317898164 and parameters: {'feature_fraction': 0.8888558017534705, 'num_leaves': 115, 'bagging_fraction': 0.9264342202471387, 'min_sum_hessian_in_leaf': 3.477856911122034}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  21%|██        | 21/100 [02:45<09:40,  7.35s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:52,597 - optuna.study.study - INFO - Trial 21 finished with value: -1189.558596227155 and parameters: {'feature_fraction': 0.6604250893099085, 'num_leaves': 28, 'bagging_fraction': 0.8481862803659174, 'min_sum_hessian_in_leaf': 0.06839576680998641}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  22%|██▏       | 22/100 [02:49<08:10,  6.28s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:15:59,415 - optuna.study.study - INFO - Trial 22 finished with value: -1185.7332203550184 and parameters: {'feature_fraction': 0.6407128341597126, 'num_leaves': 49, 'bagging_fraction': 0.7805406537287983, 'min_sum_hessian_in_leaf': 0.06225928715298572}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  23%|██▎       | 23/100 [02:56<08:16,  6.44s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:16:04,359 - optuna.study.study - INFO - Trial 23 finished with value: -1186.7806546085976 and parameters: {'feature_fraction': 0.7720767737052636, 'num_leaves': 65, 'bagging_fraction': 0.7882359189442493, 'min_sum_hessian_in_leaf': 0.23561905618095152}. Best is trial 12 with value: -1185.5412967928528.


Optimization Progress:  24%|██▍       | 24/100 [03:01<07:35,  5.99s/it, best_trial=12, best_value=-1.19e+3]

2025-12-20 21:16:15,595 - optuna.study.study - INFO - Trial 24 finished with value: -1184.3193933180985 and parameters: {'feature_fraction': 0.5646209559062487, 'num_leaves': 16, 'bagging_fraction': 0.7613976508785086, 'min_sum_hessian_in_leaf': 1.0092271935851194}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  25%|██▌       | 25/100 [03:12<09:27,  7.57s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:16:22,401 - optuna.study.study - INFO - Trial 25 finished with value: -1188.3407071212926 and parameters: {'feature_fraction': 0.5646435101985086, 'num_leaves': 18, 'bagging_fraction': 0.6700165776372807, 'min_sum_hessian_in_leaf': 1.087931568947133}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  26%|██▌       | 26/100 [03:19<09:03,  7.34s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:16:29,866 - optuna.study.study - INFO - Trial 26 finished with value: -1186.3553394551386 and parameters: {'feature_fraction': 0.5994038135058668, 'num_leaves': 153, 'bagging_fraction': 0.751503354840311, 'min_sum_hessian_in_leaf': 0.5601755973204232}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  27%|██▋       | 27/100 [03:26<08:58,  7.38s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:16:43,492 - optuna.study.study - INFO - Trial 27 finished with value: -1186.3749468583987 and parameters: {'feature_fraction': 0.5381967563388926, 'num_leaves': 171, 'bagging_fraction': 0.8951227618297071, 'min_sum_hessian_in_leaf': 1.6318941316628202}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  28%|██▊       | 28/100 [03:40<11:06,  9.25s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:16:48,935 - optuna.study.study - INFO - Trial 28 finished with value: -1188.2412262577066 and parameters: {'feature_fraction': 0.6919174275200268, 'num_leaves': 116, 'bagging_fraction': 0.7582663100218129, 'min_sum_hessian_in_leaf': 3.995971975861244}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  29%|██▉       | 29/100 [03:45<09:35,  8.11s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:16:57,272 - optuna.study.study - INFO - Trial 29 finished with value: -1186.9886800644128 and parameters: {'feature_fraction': 0.5028488995170065, 'num_leaves': 88, 'bagging_fraction': 0.8264423346529435, 'min_sum_hessian_in_leaf': 0.25069226040713516}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  30%|███       | 30/100 [03:54<09:32,  8.18s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:17:03,848 - optuna.study.study - INFO - Trial 30 finished with value: -1189.1291166124367 and parameters: {'feature_fraction': 0.6865678832536906, 'num_leaves': 207, 'bagging_fraction': 0.6897302810906769, 'min_sum_hessian_in_leaf': 0.2637360094916101}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  31%|███       | 31/100 [04:00<08:51,  7.70s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:17:08,669 - optuna.study.study - INFO - Trial 31 finished with value: -1187.147717938419 and parameters: {'feature_fraction': 0.6398773070753296, 'num_leaves': 43, 'bagging_fraction': 0.7714710482104175, 'min_sum_hessian_in_leaf': 0.10457458321357427}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  32%|███▏      | 32/100 [04:05<07:44,  6.83s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:17:15,712 - optuna.study.study - INFO - Trial 32 finished with value: -1186.9410208059553 and parameters: {'feature_fraction': 0.5736205326446837, 'num_leaves': 66, 'bagging_fraction': 0.8678789115567332, 'min_sum_hessian_in_leaf': 0.7630736537669606}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  33%|███▎      | 33/100 [04:12<07:42,  6.90s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:17:23,432 - optuna.study.study - INFO - Trial 33 finished with value: -1188.4835869716849 and parameters: {'feature_fraction': 0.6032871300830959, 'num_leaves': 16, 'bagging_fraction': 0.7395506347983078, 'min_sum_hessian_in_leaf': 0.04736300521744113}. Best is trial 24 with value: -1184.3193933180985.


Optimization Progress:  34%|███▍      | 34/100 [04:20<07:51,  7.14s/it, best_trial=24, best_value=-1.18e+3]

2025-12-20 21:17:33,949 - optuna.study.study - INFO - Trial 34 finished with value: -1184.2349944749767 and parameters: {'feature_fraction': 0.5469610259696709, 'num_leaves': 54, 'bagging_fraction': 0.6216645798963383, 'min_sum_hessian_in_leaf': 0.3306868489920368}. Best is trial 34 with value: -1184.2349944749767.


Optimization Progress:  35%|███▌      | 35/100 [04:30<08:50,  8.16s/it, best_trial=34, best_value=-1.18e+3]

2025-12-20 21:17:42,143 - optuna.study.study - INFO - Trial 35 finished with value: -1188.9035518243054 and parameters: {'feature_fraction': 0.5465258547556023, 'num_leaves': 61, 'bagging_fraction': 0.6183977902112512, 'min_sum_hessian_in_leaf': 2.1064227127039854}. Best is trial 34 with value: -1184.2349944749767.


Optimization Progress:  36%|███▌      | 36/100 [04:39<08:42,  8.17s/it, best_trial=34, best_value=-1.18e+3]

2025-12-20 21:17:47,794 - optuna.study.study - INFO - Trial 36 finished with value: -1189.2518944909639 and parameters: {'feature_fraction': 0.5762539534942337, 'num_leaves': 72, 'bagging_fraction': 0.5366707164684015, 'min_sum_hessian_in_leaf': 5.818931108952822}. Best is trial 34 with value: -1184.2349944749767.


Optimization Progress:  37%|███▋      | 37/100 [04:44<07:46,  7.41s/it, best_trial=34, best_value=-1.18e+3]

2025-12-20 21:17:52,572 - optuna.study.study - INFO - Trial 37 finished with value: -1191.5688379430649 and parameters: {'feature_fraction': 0.528768689879976, 'num_leaves': 34, 'bagging_fraction': 0.5505123381691686, 'min_sum_hessian_in_leaf': 0.17510534832564842}. Best is trial 34 with value: -1184.2349944749767.


Optimization Progress:  38%|███▊      | 38/100 [04:49<06:50,  6.62s/it, best_trial=34, best_value=-1.18e+3]

2025-12-20 21:17:57,397 - optuna.study.study - INFO - Trial 38 finished with value: -1191.3638375573819 and parameters: {'feature_fraction': 0.6794446347966296, 'num_leaves': 97, 'bagging_fraction': 0.5770641914623624, 'min_sum_hessian_in_leaf': 0.37639334881393277}. Best is trial 34 with value: -1184.2349944749767.


Optimization Progress:  39%|███▉      | 39/100 [04:54<06:11,  6.08s/it, best_trial=34, best_value=-1.18e+3]

2025-12-20 21:18:06,441 - optuna.study.study - INFO - Trial 39 finished with value: -1188.8230623593004 and parameters: {'feature_fraction': 0.6203815862613451, 'num_leaves': 121, 'bagging_fraction': 0.6351879841938949, 'min_sum_hessian_in_leaf': 1.016750244472488}. Best is trial 34 with value: -1184.2349944749767.


Optimization Progress:  40%|████      | 40/100 [05:03<07:35,  7.58s/it, best_trial=34, best_value=-1.18e+3]

[21:18:06] Hyperparameters optimization for Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM completed
[21:18:06] Start fitting Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM ...


[21:18:43] Fitting Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM finished. score = -1196.1534937834376
[21:18:43] Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM fitting and predicting completed
[21:18:43] Time left 471.49 secs

[21:18:43] Layer 2 training completed.

[21:18:43] Automl preset training completed in 428.57 seconds

[21:18:43] Model description:
Models on level 0:
	 5 averaged models Lvl_0_Pipe_0_Mod_0_LinearL2
	 5 averaged models Lvl_0_Pipe_1_Mod_0_LightGBM

Final prediction for new objects (level 1) = 
	 1.00000 * (5 averaged models Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM) 

[21:18:43] ==================================================
[21:18:43] Start 1 automl preset configuration:
[21:18:43] conf_1_sel_type_1.yml, random state: {'reader_params': {'random_state': 43}, 'nn_params': {'random_state': 43}, 'general_params': {'return_all_predictions': False}}
[21:18:43] Stdout logging level is INFO.
[21:18:43] Task: reg

[21:18:43] Start automl preset with listed constraints:
[21:18:43] - time: 471.42 sec

Optimization Progress:   0%|          | 0/100 [00:00<?, ?it/s]

2025-12-20 21:23:09,771 - optuna.storages._in_memory - INFO - A new study created in memory with name: no-name-ed3a45fa-e1ad-47d5-b01e-fe33cd504b1e
2025-12-20 21:23:19,906 - optuna.study.study - INFO - Trial 0 finished with value: -1192.9495182400349 and parameters: {'feature_fraction': 0.6872700594236812, 'num_leaves': 244, 'bagging_fraction': 0.8659969709057025, 'min_sum_hessian_in_leaf': 0.24810409748678114}. Best is trial 0 with value: -1192.9495182400349.


Optimization Progress:   1%|          | 1/100 [00:10<16:43, 10.14s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:23:27,483 - optuna.study.study - INFO - Trial 1 finished with value: -1194.1330525706153 and parameters: {'feature_fraction': 0.5780093202212182, 'num_leaves': 53, 'bagging_fraction': 0.5290418060840998, 'min_sum_hessian_in_leaf': 2.915443189153755}. Best is trial 0 with value: -1192.9495182400349.


Optimization Progress:   2%|▏         | 2/100 [00:17<14:05,  8.63s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:23:34,907 - optuna.study.study - INFO - Trial 2 finished with value: -1197.7287769193172 and parameters: {'feature_fraction': 0.8005575058716043, 'num_leaves': 185, 'bagging_fraction': 0.5102922471479012, 'min_sum_hessian_in_leaf': 7.5794799533480015}. Best is trial 0 with value: -1192.9495182400349.


Optimization Progress:   3%|▎         | 3/100 [00:25<13:03,  8.08s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:23:39,371 - optuna.study.study - INFO - Trial 3 finished with value: -1195.586114224877 and parameters: {'feature_fraction': 0.9162213204002109, 'num_leaves': 66, 'bagging_fraction': 0.5909124836035503, 'min_sum_hessian_in_leaf': 0.00541524411940254}. Best is trial 0 with value: -1192.9495182400349.


Optimization Progress:   4%|▍         | 4/100 [00:29<10:38,  6.65s/it, best_trial=0, best_value=-1.19e+3]

2025-12-20 21:23:50,802 - optuna.study.study - INFO - Trial 4 finished with value: -1189.5368790958246 and parameters: {'feature_fraction': 0.6521211214797689, 'num_leaves': 141, 'bagging_fraction': 0.7159725093210578, 'min_sum_hessian_in_leaf': 0.014618962793704969}. Best is trial 4 with value: -1189.5368790958246.


Optimization Progress:   5%|▌         | 5/100 [00:41<13:15,  8.38s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:23:53,829 - optuna.study.study - INFO - Trial 5 finished with value: -1197.7644325152635 and parameters: {'feature_fraction': 0.8059264473611898, 'num_leaves': 49, 'bagging_fraction': 0.6460723242676091, 'min_sum_hessian_in_leaf': 0.029204338471814112}. Best is trial 4 with value: -1189.5368790958246.


Optimization Progress:   6%|▌         | 6/100 [00:44<10:16,  6.56s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:24:00,730 - optuna.study.study - INFO - Trial 6 finished with value: -1197.8451378692523 and parameters: {'feature_fraction': 0.728034992108518, 'num_leaves': 204, 'bagging_fraction': 0.5998368910791798, 'min_sum_hessian_in_leaf': 0.11400863701127326}. Best is trial 4 with value: -1189.5368790958246.


Optimization Progress:   7%|▋         | 7/100 [00:50<10:20,  6.67s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:24:07,132 - optuna.study.study - INFO - Trial 7 finished with value: -1192.5318303291822 and parameters: {'feature_fraction': 0.7962072844310213, 'num_leaves': 27, 'bagging_fraction': 0.8037724259507192, 'min_sum_hessian_in_leaf': 0.004809461967501573}. Best is trial 4 with value: -1189.5368790958246.


Optimization Progress:   8%|▊         | 8/100 [00:57<10:05,  6.58s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:24:19,095 - optuna.study.study - INFO - Trial 8 finished with value: -1191.5014420024315 and parameters: {'feature_fraction': 0.5325257964926398, 'num_leaves': 243, 'bagging_fraction': 0.9828160165372797, 'min_sum_hessian_in_leaf': 1.7123375973163983}. Best is trial 4 with value: -1189.5368790958246.


Optimization Progress:   9%|▉         | 9/100 [01:09<12:32,  8.27s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:24:27,276 - optuna.study.study - INFO - Trial 9 finished with value: -1189.5901198541906 and parameters: {'feature_fraction': 0.6523068845866853, 'num_leaves': 39, 'bagging_fraction': 0.8421165132560784, 'min_sum_hessian_in_leaf': 0.057624872164786005}. Best is trial 4 with value: -1189.5368790958246.


Optimization Progress:  10%|█         | 10/100 [01:17<12:21,  8.24s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:24:37,110 - optuna.study.study - INFO - Trial 10 finished with value: -1192.5035548818325 and parameters: {'feature_fraction': 0.9725682721151933, 'num_leaves': 109, 'bagging_fraction': 0.7149885992524332, 'min_sum_hessian_in_leaf': 0.0011799062523159282}. Best is trial 4 with value: -1189.5368790958246.


Optimization Progress:  11%|█         | 11/100 [01:27<12:56,  8.73s/it, best_trial=4, best_value=-1.19e+3]

2025-12-20 21:24:49,957 - optuna.study.study - INFO - Trial 11 finished with value: -1188.6421801264853 and parameters: {'feature_fraction': 0.6075747423853284, 'num_leaves': 125, 'bagging_fraction': 0.871545830150035, 'min_sum_hessian_in_leaf': 0.02616865856724872}. Best is trial 11 with value: -1188.6421801264853.


Optimization Progress:  12%|█▏        | 12/100 [01:40<12:14,  8.35s/it, best_trial=11, best_value=-1.19e+3]

[21:24:49] Hyperparameters optimization for Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM completed
[21:24:49] Start fitting Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM ...


[21:25:26] Fitting Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM finished. score = -1193.86867307248
[21:25:26] Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM fitting and predicting completed
[21:25:26] Time left 68.21 secs

[21:25:26] Layer 2 training completed.

[21:25:26] Automl preset training completed in 204.52 seconds

[21:25:26] Model description:
Models on level 0:
	 5 averaged models Lvl_0_Pipe_0_Mod_0_LinearL2
	 5 averaged models Lvl_0_Pipe_1_Mod_0_LightGBM

Final prediction for new objects (level 1) = 
	 1.00000 * (5 averaged models Lvl_1_Pipe_0_Mod_0_Tuned_LightGBM) 

[21:25:26] ==================================================
[21:25:26] Blending: optimization starts with equal weights. Score = -1282.7782150
[21:25:26] Blending: iteration 0: score = -1186.8110449, weights = [0.4724345 0.        0.        0.        0.5275655]
[21:25:26] Blending: no improvements for score. Terminated.

[21:25:26] Blending: best score = -1186.8110449, best weights = [0.4724345 0.        0.        0.        0.52756

## Сравнение результатов
Сравним метрики качества двух конфигураций LAMA на кросс-валидации.


In [5]:
results = pd.DataFrame({
    'Configuration': ['Default', 'Tuned'],
    'Log MAE': [metrics_default['log_mae'], metrics_tuned['log_mae']],
})

print(results.to_string(index=False))

best_idx = results['Log MAE'].idxmin()
best_config = results.loc[best_idx, 'Configuration']
best_log_mae = results.loc[best_idx, 'Log MAE']

print(f"Best LAMA Configuration: {best_config}")

Configuration  Log MAE
      Default 0.523108
        Tuned 0.436323
Best LAMA Configuration: Tuned


## Генерация предсказаний на тестовой выборке

Используем лучшую модель для генерации предсказаний на тестовых данных


In [6]:
best_model = lama_default if metrics_default['log_mae'] < metrics_tuned['log_mae'] else lama_tuned

test_predictions = best_model.predict(test_df)

print(f"Predictions statistics:")
print(f"Mean: {test_predictions.mean():.2f}")
print(f"Std: {test_predictions.std():.2f}")
print(f"Min: {test_predictions.min():.2f}")
print(f"Max: {test_predictions.max():.2f}")

baseline_log_mae = best_log_mae
print(f"LAMA Baseline Log MAE to beat: {baseline_log_mae:.4f}")


Predictions statistics:
Mean: 3032.10
Std: 2172.39
Min: 585.96
Max: 39201.54
LAMA Baseline Log MAE to beat: 0.4363


## Выводы и анализ результатов

### Результаты экспериментов

| Конфигурация | Log MAE | Raw MAE | Улучшение |
|--------------|---------|---------|-----------|
| Default      | 0.5231  | ~1297   | —         |
| Tuned        | 0.4363  | ~1187   | **16.6%** |

### Ключевые наблюдения

1. В базовой конфигурации Linear L2 получила вес ~93.5% в финальном блендинге - это говорит о том, что линейная модель хорошо справляется с данной задаче. LightGBM добавляет лишь небольшое улучшение на первом уровне
2. Запуск нескольких конфигураций с разными `random_state` обеспечивает более робастные результаты

### Следующие шаги: Custom Solution (`03_custom_solution.ipynb`)

В следующем ноутбуке реализуем кастомное решение для сравнения с LAMA baseline:

1. **Три независимых пайплайна**: LightGBM, XGBoost и CatBoost с логарифмированием таргета
2. **Optuna-оптимизация**: автоматический подбор гиперпараметров LightGBM с оптимизацией Log MAE
3. **Взвешенный ансамбль**: объединение предсказаний моделей с весами, обратно пропорциональными их ошибкам
4. **EDA-инсайты**: использование рекомендаций из `01_eda.ipynb` (логарифмирование таргета, топовые MI-фичи)
